### PR TITLE
[Enhancement] Task framework support merge task run

### LINKED ISSUE #8231: Task framework support merge task run
## Enhancement
For materialized view refresh, users may use periodic refresh to cause a large number of TaskRuns to accumulate. For this kind of TaskRuns that are automatically scheduled by the system, we should have a combined strategy to reduce unnecessary refreshes.

### PR DESCRIPTION
**Pull request by Astralidea**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #8231

## Problem Summary(Required) ：

 At present, only the manual and automatic tasks of the materialized view have different priorities.The manual priority is higher. For manual tasks, we do not merge operations. For automatic tasks, we will compare the definition, and if they are the same, we will perform the merge operation.

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 42 / 58 (72.41%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunManager.java|21|31|67.74%|
|:large_blue_circle:|com/starrocks/scheduler/ExecuteOption.java|9|13|69.23%|
|:large_blue_circle:|com/starrocks/scheduler/persist/TaskRunStatus.java|3|4|75.00%|
|:large_blue_circle:|com/starrocks/scheduler/TaskManager.java|8|9|88.89%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|1|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/scheduler/TaskRunManager.java [hunk 1]
```diff
@@ -57,16 +61,40 @@ public SubmitResult submitTaskRun(TaskRun taskRun, int priority) {
...
+        if (!tryTaskRunLock()) {
+            return;
+        }
+        try {
+            long taskId = taskRun.getTaskId();
+            MinMaxPriorityQueue<TaskRun> taskRuns = pendingTaskRunMap.computeIfAbsent(taskId,
+                    u -> MinMaxPriorityQueue.orderedBy(TaskRun::compareTo).create());
+            if (mergeRedundant && taskRuns.size() >= 1) {
+                TaskRun lastTaskRun = taskRuns.peekLast();
+                if (lastTaskRun.getStatus().getDefinition().equals(taskRun.getStatus().getDefinition())) {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/scheduler/TaskRunManager.java:85]: 1. The definition between task runs is different?
2. Why only support to merge with the last one? If the definition is the same, can it be merged with the task run in the middle?

File: fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java [hunk 2]
```diff
@@ -3015,7 +3016,7 @@ public void refreshMaterializedView(String dbName, String mvName, int priority)
             Task task = TaskBuilder.buildMvTask(materializedView, dbName);
             taskManager.createTask(task, false);
         }
-        taskManager.executeTask(mvTaskName, priority);
+        taskManager.executeTask(mvTaskName, new ExecuteOption(priority));
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:3019]: it seems that mergeRedundant is default false. It means you disable merge by default?

**Review by ABingHuang — CHANGES_REQUESTED** [review]

File: fe/fe-core/src/main/java/com/starrocks/scheduler/TaskRunManager.java (hunk 1 above)
Comment (Astralidea) [reply fe/fe-core/src/main/java/com/starrocks/scheduler/TaskRunManager.java]: 1. Yes, For materialized view scenarios, there may be some partition refreshes that lead to different definitions.
2. You are right. I use another way to implement it .

File: fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java (hunk 2 above)
Comment (Astralidea) [reply fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:3019]: Yes. For the periodical tasks generated by our system, I think merge can be turned on, and mrege should not be done by default for user trigger.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 43 / 62 (69.35%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/scheduler/TaskRun.java|5|9|55.56%|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunManager.java|17|26|65.38%|
|:large_blue_circle:|com/starrocks/scheduler/ExecuteOption.java|9|13|69.23%|
|:large_blue_circle:|com/starrocks/scheduler/persist/TaskRunStatus.java|3|4|75.00%|
|:large_blue_circle:|com/starrocks/scheduler/TaskManager.java|8|9|88.89%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|1|100.00%|

**Review by ABingHuang — DISMISSED** [review]
LGTM

File: fe/fe-core/src/main/java/com/starrocks/scheduler/TaskRunManager.java [hunk 3]
```diff
@@ -57,16 +62,37 @@ public SubmitResult submitTaskRun(TaskRun taskRun, int priority) {
...
+            return;
+        }
+        try {
+            long taskId = taskRun.getTaskId();
+            PriorityBlockingQueue<TaskRun> taskRuns = pendingTaskRunMap.computeIfAbsent(taskId,
+                    u -> Queues.newPriorityBlockingQueue());
+            if (mergeRedundant) {
+                // The remove here is actually remove the old TaskRun.
+                // Note that the old TaskRun and new TaskRun may have the same priority and definition,
+                // but other attributes may be different, such as creation time.
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/scheduler/TaskRunManager.java:87]: Why don't keep only one task run with the highest priority if task runs have the same definition?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 53 / 75 (70.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/scheduler/TaskRun.java|4|7|57.14%|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunManager.java|28|41|68.29%|
|:large_blue_circle:|com/starrocks/scheduler/ExecuteOption.java|9|13|69.23%|
|:large_blue_circle:|com/starrocks/scheduler/persist/TaskRunStatus.java|3|4|75.00%|
|:large_blue_circle:|com/starrocks/scheduler/TaskManager.java|8|9|88.89%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 55 / 77 (71.43%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/scheduler/TaskRun.java|4|7|57.14%|
|:large_blue_circle:|com/starrocks/scheduler/TaskRunManager.java|30|43|69.77%|
|:large_blue_circle:|com/starrocks/scheduler/ExecuteOption.java|9|13|69.23%|
|:large_blue_circle:|com/starrocks/scheduler/persist/TaskRunStatus.java|3|4|75.00%|
|:large_blue_circle:|com/starrocks/scheduler/TaskManager.java|8|9|88.89%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|1|100.00%|

**Review by wyb — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/scheduler/TaskRunManager.java [hunk 4]
```diff
@@ -57,16 +63,66 @@ public SubmitResult submitTaskRun(TaskRun taskRun, int priority) {
...
+    }
+
+    // Because java PriorityQueue does not provide an interface for searching by element,
+    // so find it by code O(n), which can be optimized later
+    @Nullable
+    private TaskRun getTaskRun(PriorityBlockingQueue<TaskRun> taskRuns, TaskRun taskRun) {
+        TaskRun oldTaskRun = null;
+        for (TaskRun run : taskRuns) {
+            if (run.equals(taskRun)) {
+                oldTaskRun = run;
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/scheduler/TaskRunManager.java:119]: return run;
oldTaskRun is useless, just return.

**Review by ABingHuang — APPROVED** [review]
LGTM

**Review by chaoyli — APPROVED** [review]

**Pull request merged**